### PR TITLE
Monaco: add suggestions for template variables

### PR DESCRIPTION
**Pull request by ryantxu**
WIP... posting here for early feedback

With many of the code editors, we need to support a standard language *and* apply template variables.  For example the text panel is HTML/markdown with and dashboard templates applied.  SQL etc, is standard SQL + a set of $__macro functions we define

image

### CODE REVIEW

File: packages/grafana-ui/src/components/Monaco/CodeEditor.tsx [hunk 1]
```diff
@@ -1,35 +1,113 @@
...
+      insertText: '${__field.name}',
+      range: range,
+      detail: 'Keword detail',
+    },
+  ];
+}
+function registerXXXX(language: string): monaco.IDisposable {
+  console.log('registerCompletionItemProvider', language);
+  return monaco.languages.registerCompletionItemProvider(language, {
+    triggerCharacters: ['$'],
```
Comment (ryantxu) [inline packages/grafana-ui/src/components/Monaco/CodeEditor.tsx:56]: Special key to open with `$`

File: packages/grafana-ui/src/components/Monaco/CodeEditor.tsx [hunk 2]
```diff
@@ -1,35 +1,113 @@
...
+      documentation: 'The field value name (...)',
+      insertText: '${__field.name}',
+      range: range,
+      detail: 'Keword detail',
+    },
+  ];
+}
+function registerXXXX(language: string): monaco.IDisposable {
+  console.log('registerCompletionItemProvider', language);
+  return monaco.languages.registerCompletionItemProvider(language, {
```
Comment (ryantxu) [inline packages/grafana-ui/src/components/Monaco/CodeEditor.tsx:55]: sadly -- it looks like completion needs to be added globally to the language :(  would be much nicer if it was linked to the editor instance.

It does look like you can register and destory the instances, but this has the not great side affect that really one suggest context is active for each language at a time.

NOTE: this works on top of the standard language completions :tada:

File: packages/grafana-ui/src/components/Monaco/types.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  Method = 0,
+  Function = 1,
+  Field = 3,
+  Variable = 4,
+  Property = 9,
+  Constant = 14,
+  Keyword = 17,
+}
+
+export interface CodeEditorSuggestionItem {
```
Comment (ryantxu) [inline packages/grafana-ui/src/components/Monaco/types.ts:41]: Very similar shape to https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/types/completion.ts#L34

File: packages/grafana-ui/src/components/Monaco/CodeEditor.tsx (hunk 1 above)
Comment (torkelo) [reply packages/grafana-ui/src/components/Monaco/CodeEditor.tsx:56]: nice!

**Comment by ryantxu** [comment]
@torkelo @dprokop -- this would be really great to get into beta1.  It solves key issues for timestream :)

You can test using the text panel on a page with template variables:
image

File: packages/grafana-ui/src/components/Monaco/types.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
+export type CodeEditorChangeHandler = (value: string) => void;
+export type CodeEditorSuggestionProvider = () => CodeEditorSuggestionItem[];
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Monaco/types.ts:2]: This type feels too simplistic, needs prefix & context so suggestions can be context specific. So we will have to break this api later

File: packages/grafana-ui/src/components/Monaco/utils.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { VariableSuggestion } from '@grafana/data';
+import { CodeEditorSuggestionItem, CodeEditorSuggestionItemKind } from './types';
+
+/**
+ * @alpha
+ */
+export function variableSuggestionToCodeEditorSuggestion(sug: VariableSuggestion): CodeEditorSuggestionItem {
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Monaco/utils.ts:7]: Not a huge fan of this mapping. I mean we could probably try to unify the VariablesSuggestion and CodeEditorSuggestionItem to have one interface. I'll try to do that.

File: packages/grafana-ui/src/components/Monaco/types.ts (hunk 4 above)
Comment (ryantxu) [reply packages/grafana-ui/src/components/Monaco/types.ts:2]: For complex things, shouldn't we just use monaco/lsp?  The goal here is to have simple support for the super common templates/variables.

Complex things can be supported by either loading custom languages:
https://github.com/microsoft/monaco-languages/tree/master/src

or customizing in JS:
https://microsoft.github.io/monaco-editor/playground.html#extending-language-services-custom-languages

Comment (ryantxu) [reply packages/grafana-ui/src/components/Monaco/types.ts:2]: Note that these suggestions exist on top of the defined language.

File: packages/grafana-ui/src/components/Monaco/types.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  /**
+   * A human-readable string that represents a doc-comment.
+   */
+  documentation?: string; // | IMarkdownString;
+
+  /**
+   * A string or snippet that should be inserted in a document when selecting
+   * this completion. When `falsy` the `label` is used.
+   */
+  insertText?: string;
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Monaco/types.ts:76]: The label vs insertText. I think we can make the insertText not to be optional, to be more in line with label/value in VariableSuggestion interface. But then there's https://github.com/grafana/grafana/pull/25921/files#diff-f02ee9fdd9a1eb7c16cb24b321ddee2aR12 that assumes `$` for suggestions retrieval and it's going through the labels to find the matching suggestion. Think we need to update this to take the insert text into consideration when generating the suggestions list.

**Review by dprokop — APPROVED** [review]

**Pull request merged**